app/src/main/java/zapsolutions/zap/transactionHistory/transactionDetails/InvoiceDetailBSDFragment.java

```diff
@@ -0,0 +1,221 @@
+                if (Wallet.getInstance().isInvoiceExpired(invoice)) {
+                    bindExpiredInvoice(invoice);
+                } else {
+                    // The invoice has not yet expired
+                    bindOpenInvoice(invoice);
+                }
+            }
+        } else {
+            // if a specific value was requested
+            if (amtPayed.equals(amt)) {
+                // The invoice has been payed
+                bindPayedInvoice(invoice);
+            } else {
+                // The invoice has not been payed yet
+                if (Wallet.getInstance().isInvoiceExpired(invoice)) {
+                    // The invoice has expired
+                    bindExpiredInvoice(invoice);
+                } else {
+                    // The invoice has not yet expired
+                    bindOpenInvoice(invoice);
+                }
+            }
+        }
+    }
+
+    private void bindOpenInvoice(Invoice invoice) {
+        mTransactionDescription.setText(R.string.invoice_detail);
+
+        mAmount.setText(MonetaryUtil.getInstance().getPrimaryDisplayAmountAndUnit(invoice.getValue()));
+
+        // Generate "QR-Code"
+        Bitmap bmpQRCode = QRCode
+                .from(invoice.getPaymentRequest())
+                .withSize(500, 500)
+                .withErrorCorrection(ErrorCorrectionLevel.L)
+                .bitmap();
+        mQRCodeView.setImageBitmap(bmpQRCode);
+
+        ScheduledExecutorService expiryUpdateSchedule =
+                Executors.newSingleThreadScheduledExecutor();
+
+        expiryUpdateSchedule.scheduleAtFixedRate
+                (new Runnable() {
+                    public void run() {
+                        long timeLeft = (invoice.getCreationDate() + invoice.getExpiry()) - (System.currentTimeMillis() / 1000);
+                        String expiryText = formattedDuration(timeLeft) + " " + getActivity().getResources().getString(R.string.remaining);
+
+                        mExpiry.setText(expiryText);
+                    }
+                }, 0, 1, TimeUnit.SECONDS);
+    }
+
+    private void bindPayedInvoice(Invoice invoice) {
+        mTransactionDescription.setText(R.string.transaction_detail);
+        mExpiryLabel.setVisibility(View.GONE);
+        mExpiry.setVisibility(View.GONE);
+        mAmount.setText(MonetaryUtil.getInstance().getPrimaryDisplayAmountAndUnit(invoice.getAmtPaidSat()));
+        mQRCodeView.setVisibility(View.GONE);
+    }
+
+    private void bindExpiredInvoice(Invoice invoice) {
+        mTransactionDescription.setText(R.string.invoice_detail);
+        mAmount.setText(MonetaryUtil.getInstance().getPrimaryDisplayAmountAndUnit(invoice.getValue()));
+        mExpiry.setText(R.string.expired);
+        mQRCodeView.setVisibility(View.GONE);
+    }
+
+    /**
+     * Returns a nicely formatted duration.
+     *
+     * @param duration in seconds
+     * @return
+     */
+    private String formattedDuration(long duration) {
```

Comment:
Wonder if we should move this to a util class, as it could be used for other cases as well in the future.